sw/device/silicon_creator/lib/drivers/otp.c

```diff
@@ -0,0 +1,77 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
+
+#include <stddef.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "otp_ctrl_regs.h"
+
+static rom_error_t otp_read_start(const otp_t *otp, uint32_t address) {
+  uint32_t status =
+      mmio_region_read32(otp->base_addr, OTP_CTRL_STATUS_REG_OFFSET);
+  // Check if otp_ctrl is busy.
+  if (!bitfield_bit32_read(status, OTP_CTRL_STATUS_DAI_IDLE_BIT)) {
+    return kErrorOtpBusy;
+  }
+
+  // Issue read command.
+  uint32_t cmd =
+      mmio_region_read32(otp->base_addr, OTP_CTRL_DIRECT_ACCESS_CMD_REG_OFFSET);
+  cmd = bitfield_bit32_write(cmd, OTP_CTRL_DIRECT_ACCESS_CMD_RD_BIT, 1u);
+  mmio_region_write32(otp->base_addr, OTP_CTRL_DIRECT_ACCESS_CMD_REG_OFFSET,
+                      cmd);
+
+  // Wait for read to complete.
+  do {
+    status = mmio_region_read32(otp->base_addr, OTP_CTRL_STATUS_REG_OFFSET);
+  } while (!bitfield_bit32_read(status, OTP_CTRL_STATUS_DAI_IDLE_BIT));
+
+  return kErrorOk;
+}
+
+uint32_t otp_read32(const otp_t *otp, uint32_t address) {
+  return mmio_region_read32(otp->base_addr,
+                            OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address);
+}
+uint64_t otp_read64(const otp_t *otp, uint32_t address) {
+  ptrdiff_t reg_offset = OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address;
+  uint32_t w1 = mmio_region_read32(otp->base_addr, reg_offset);
+  uint32_t w2 =
```

Comment:
I'd make these both u64 to avoid the cast below and any concerns around promotion.